File: Algorithmic skills/firstUnique.cpp
// The problem: https://app.codility.com/programmers/trainings/4/first_unique/
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

int firstUmique(vector<int> &A)
{
  // mapping the values and there appearance
  unordered_map<int, int> umap;
  for (int i = 0; i < A.size(); i++)
  {
    if (umap.find(A[i]) == umap.end())
    {
      umap[A[i]] = 1;
    }
    else
    {
      // if it already appears in the array, give it a unique count
      umap[A[i]] = 5;
    }
  }
  // return the first element with that aappears only once
  for (int i = 0; i < A.size(); i++)
  {
    if (umap[A[i]] == 1)
    {
      return A[i];
    }
  }
  return -1;
};

int main()
{
  vector<int> v = {10, 13, 16, 16, 18};
  cout << firstUmique(v) << endl;
}

File: Algorithmic skills/first_unique.py
# The problem: https://app.codility.com/programmers/trainings/4/first_unique/
from collections import Counter


def first_unique(A):
    # Counting the items, keep only those which have a count of 1:
    all_unique = [k for k, v in Counter(A).items() if v == 1]
    if len(all_unique) == 0:
        return -1
    return all_unique[0]


if __name__ == '__main__':
    A = [2, 2, 3, 4, 5, 5]
    print(first_unique(A))
